engine/src/main/java/org/terasology/network/internal/ServerConnectListManager.java

```diff
@@ -0,0 +1,146 @@
+ * limitations under the License.
+ */
+package org.terasology.network.internal;
+
+import com.google.gson.Gson;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.terasology.context.Context;
+import org.terasology.engine.paths.PathManager;
+import org.terasology.engine.subsystem.DisplayDevice;
+import org.terasology.i18n.TranslationSystem;
+
+import java.io.IOException;
+import java.io.Writer;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Collections;
+import java.util.Set;
+
+/**
+ * This class provides the methods needed to determine if a client is allowed to connect or not,
+ * based on the blacklist and whitelist files.
+ */
+
+public class ServerConnectListManager {
+
+    private static final Logger logger = LoggerFactory.getLogger(ServerConnectListManager.class);
+    private static final Gson gson = new Gson();
+
+    private Context context;
+    private Set blacklistedIDs;
+    private Set whitelistedIDs;
+    private final Path blacklistPath;
+    private final Path whitelistPath;
+
+    public ServerConnectListManager(Context context) {
+        blacklistPath = PathManager.getInstance().getHomePath().resolve("blacklist.json");
+        whitelistPath = PathManager.getInstance().getHomePath().resolve("whitelist.json");
+        this.context = context;
+        loadLists();
+    }
+
+    private void loadLists() {
+        try {
+            if (createFiles()) {
+                blacklistedIDs = gson.fromJson(Files.newBufferedReader(blacklistPath), Set.class);
+                whitelistedIDs = gson.fromJson(Files.newBufferedReader(whitelistPath), Set.class);
+            }
+        } catch (IOException e) {
+            logger.error("Whitelist or blacklist files not found:", e);
+        }
+    }
+
+    private void saveLists() {
+        try {
+            if (createFiles()) {
+                Writer blacklistWriter = Files.newBufferedWriter(blacklistPath);
+                Writer whitelistWriter = Files.newBufferedWriter(whitelistPath);
+                blacklistWriter.write(gson.toJson(blacklistedIDs));
+                whitelistWriter.write(gson.toJson(whitelistedIDs));
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
+
+    private boolean createFiles() throws IOException {
+        DisplayDevice display = context.get(DisplayDevice.class);
+        if (display == null || !display.isHeadless()) {
+            return false;
+        }
+        if (!Files.exists(blacklistPath)) {
+            Files.createFile(blacklistPath);
+        }
+        if (!Files.exists(whitelistPath)) {
+            Files.createFile(whitelistPath);
+        }
+        return true;
+    }
+
+    public String getErrorMessage(String clientID) {
+        TranslationSystem translationSystem = context.get(TranslationSystem.class);
+        if (isClientBlacklisted(clientID)) {
+            return translationSystem.translate("${engine:menu#client-on-blacklist}");
```

Comment:
If I understand it correctly, this way the string is translated *by the server*, but the language set in the client and the one set in the server may differ so the translation in my understanding would be quite pointless, and possibly confusing if for example the user has their client set in English but the server is set to another language the user doesn't understand.

For now I would just send the English string (basically I'd reverse the changes you made to access `TranslationSystem`, including the part which brings here the `Context`), then if there is time / anybody else wants to do it, we could work on a separate "improve localization" PR which makes the translation on the client (the "non-trivial" stuff we were discussing yesterday)

Reply:
I think you are right about how the server would translate the string based on its own settings. I suppose that when the time comes to do the translation for ClientConnectionHandler, we could change the value of getErrorMessage to return a string such as "client-on-blacklist" and use that with the translationSystem on the client side.

I still need the context to get the display so that the code will not generate blacklist and whitelist files whenever a game client starts up.

Reply:
Ah okay, I didn't notice the context is also needed for the display